hail/python/hail/expr/functions.py

```diff
@@ -7070,6 +7070,106 @@ def coerce_endpoint(point):
     )
 
 
+@typecheck(path=builtins.str, point_or_interval=expr_any, entries_name=builtins.str)
+def query_matrix_table(path, point_or_interval, entries_name='entries_array'):
+    """Query row records from a matrix table corresponding to a given point or
+    range of row keys. The entry fields are localized as an array of structs as
+    in :meth:`.MatrixTable.localize_entries`.
+
+    Notes
+    -----
+    This function does not dispatch to a distributed runtime; it can be used inside
+    already-distributed queries such as in :meth:`.Table.annotate`.
+
+    Warning
+    -------
+    This function contains no safeguards against reading large amounts of data
+    using a single thread.
+
+    Parameters
+    ----------
+    path : :class:`str`
+        Table path.
+    point_or_interval
+        Point or interval to query.
+    entries_name : :class:`str`
+        Identifier to use for the localized entries array. Must not conflict
+        with any row field identifiers.
+
+    Returns
+    -------
+    :class:`.ArrayExpression`
+    """
+    matrix_table = hl.read_matrix_table(path)
+    if entries_name in matrix_table.row:
+        raise ValueError(
+            f'query_matrix_table: field "{entries_name}" is present in matrix table row fields, pick a different `entries_name`'
+        )
+
+    entries_table = hl.read_table(path + '/entries')
+    [entry_id] = list(entries_table.row)
+
+    full_row_type = tstruct(**matrix_table.row.dtype, **entries_table.row.dtype)
+    key_typ = matrix_table.row_key.dtype
+    key_names = list(key_typ)
+    len = builtins.len
+
+    def coerce_endpoint(point):
+        if point.dtype == key_typ[0]:
+            point = hl.struct(**{key_names[0]: point})
+        ts = point.dtype
+        if isinstance(ts, tstruct):
+            i = 0
+            while i < len(ts):
+                if i >= len(key_typ):
+                    raise ValueError(
+                        f"query_matrix_table: queried with {len(ts)} row key field(s), but matrix table only has {len(key_typ)} row key field(s)"
+                    )
+                if key_typ[i] != ts[i]:
+                    raise ValueError(
+                        f"query_matrix_table: mismatch at row key field {i} ({list(ts.keys())[i]!r}): query type is {ts[i]}, matrix table row key type is {key_typ[i]}"
+                    )
+                i += 1
+
+            if i == 0:
+                raise ValueError("query_matrix_table: cannot query with empty row key")
```

Comment:
If this is just doing validation, can you pull this out into a procedure that does just that - it takes up a lot of space that doesn't really convey much in terms of data coercion.